File: Nov2020/23Nov/LinkedListComponents.java
/*
Problem Name: 817. Linked List Components
Problem Link: https://leetcode.com/problems/linked-list-components/
*/
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public int numComponents(ListNode head, int[] G) {
        Set<Integer> set = new HashSet<Integer>();
        for(int i: G)
            set.add(i);
        int cnt = 0, cur = 0;
        while(head != null){
            if(set.contains(head.val))
                cur++;
            else{
                if(cur > 0)
                    cnt++;
                cur = 0;
            }
            head = head.next;
        }
        if(cur > 0)
            cnt++;
        return cnt;
    }
}

File: Nov2020/23Nov/LinkedListComponents.cpp
/*
Problem name: 817. Linked List Components
Problem link: https://leetcode.com/problems/linked-list-components/
*/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    int numComponents(ListNode* head, vector<int>& G) {
        int count=0;
        ListNode* curr=head;
        
        set<int>s(G.begin(),G.end());
        
        while(curr!=nullptr){
            if(s.count(curr->val)){
                count++;
                while(curr!=nullptr && s.count(curr->val)){
                    curr=curr->next;
                }
            }
            else{
                curr=curr->next;
            }
        }
        return count;
    }
};
